### PR TITLE
A0-3894: A0-3891 A0-3892 Added pallet-balances-maintenance.py

### PR DESCRIPTION
**Pull request by Marcin-Radecki**
# Description

Script for maintenance operations on AlephNode chain with regard to pallet balances.

It has the following functionality: 
* workaround for bug https://github.com/paritytech/polkadot-sdk/pull/2700/files, that make sure all 
  accounts have at least ED as their free balance,
* programmatic support for sending Balances.UpgradeAccounts ext for all accounts,
* checking pallet balances and account reference counters invariants.

By default, it connects to an AlephZero Testnet and performs sanity checks only ie not changing state of the chain at all.
Accounts that do not satisfy those checks are written to `accounts-with-failed-invariants.json file`.

## Type of change

Please delete options that are not relevant.

- New feature (non-breaking change which adds functionality)

## Testing

1) Base functionality (only check invariants for all accounts), regardless 11 or 12 version:

`./pallet-balances-maintenance.py --ws-url <test-chain-url>`

```
2024-01-22 13:01:55,530 [INFO] Connected to Aleph Node: Aleph Zero Development 0.11.4-b7e984c756d
2024-01-22 13:01:55,591 [INFO] Major version of chain connected to is ChainMajorVersion.PRE_12_MAJOR_VERSION
2024-01-22 13:01:56,366 [INFO] Existential deposit is 5e-10 DZERO
2024-01-22 13:01:56,366 [INFO] Performing pallet balances sanity checks.
2024-01-22 13:01:58,548 [INFO] Checked 5000 accounts
[...]
2024-01-22 13:03:02,316 [INFO] Checked 190000 accounts
2024-01-22 13:03:02,976 [INFO] Total accounts that match given predicate 'account invariants' is 192885
2024-01-22 13:03:03,280 [INFO] All accounts on chain Aleph Zero Development meet balances invariants.
2024-01-22 13:03:03,280 [INFO] DONE
```

2) On 11 version, fixing free balance for accounts that would be invalid in 12 version

`./pallet-balances-maintenance.py --ws-url <test-chain-url> --fix-free-balance`

```
2024-01-22 13:03:46,232 [INFO] Connected to Aleph Node: Aleph Zero Development 0.11.4-b7e984c756d
2024-01-22 13:03:46,292 [INFO] Major version of chain connected to is ChainMajorVersion.PRE_12_MAJOR_VERSION
2024-01-22 13:03:47,044 [INFO] Existential deposit is 5e-10 DZERO
2024-01-22 13:03:47,050 [INFO] Using following account for transfers: 5D34dL5prEUaGNQtPPZ3yN5Y6BnkfXunKXXz6fo7ZJbLwRRH
2024-01-22 13:03:47,050 [INFO] Will send at most 64 transfers in a batch.
2024-01-22 13:03:47,051 [INFO] Looking for accounts that would be dust in 12 version.
2024-01-22 13:03:49,350 [INFO] Checked 5000 accounts
[...]
2024-01-22 13:04:53,972 [INFO] Checked 190000 accounts
2024-01-22 13:04:54,612 [INFO] Total accounts that match the given predicate 'account valid in pre-12 version but not in 12 version' is 192885
2024-01-22 13:04:54,924 [INFO] Found 510 accounts that will be invalid in 12 version.
2024-01-22 13:04:54,929 [INFO] Wrote file 'dust-accounts.json'
2024-01-22 13:04:54,929 [INFO] Adjusting balances by sending transfers.
[...]
2024-01-22 13:05:51,230 [INFO] Extrinsic included in block 0x7e592baddf44c2b9d8f19fb35c5882aeacf64870159b799a4e3aafec9d973824: Paid 0.011088659408 DZERO
2024-01-22 13:05:51,230 [INFO] Transfers done.
```

3) on 12 version, upgrading all accounts

`/pallet-balances-maintenance.py --ws-url <test-chain-url> --upgrade-accounts --upgrade-accounts-in-batch 1024`

```
2024-01-22 13:30:12,608 [INFO] Connected to Aleph Node: Aleph Zero Development 0.11.4-b7e984c756d
2024-01-22 13:30:12,666 [INFO] Major version of chain connected to is ChainMajorVersion.AT_LEAST_12_MAJOR_VERSION
2024-01-22 13:30:13,450 [INFO] Existential deposit is 5e-10 DZERO
2024-01-22 13:30:13,456 [INFO] Using following account for upgrade_accounts: 5D34dL5prEUaGNQtPPZ3yN5Y6BnkfXunKXXz6fo7ZJbLwRRH
2024-01-22 13:30:13,457 [INFO] Will upgrade at most 1024 accounts in a batch.
2024-01-22 13:30:13,457 [INFO] Querying all accounts.
2024-01-22 13:30:16,667 [INFO] Checked 5000 accounts
[...]
2024-01-22 13:31:26,236 [INFO] Checked 190000 accounts
2024-01-22 13:31:27,177 [INFO] Total accounts that match given predicate 'all accounts' is 193500
2024-01-22 13:31:27,917 [INFO] About to upgrade 1024 accounts, each with 5e-10 DZERO
2024-01-22 13:31:28,619 [INFO] Extrinsic included in block 0x37b61a0ad94b5971dbb6a4439aeadf883c8e6a63aab4eb44874ea2aac51ea69a: Paid 0.0 DZERO
[...]
```

### CODE REVIEW

**Review by DamianStraszak — APPROVED** [review]

File: scripts/accounts-invariants/pallet-balances-maintenance.py [hunk 1]
```diff
@@ -0,0 +1,451 @@
...
+    Filters our accounts from the list on which predicate returns False
+    :param chain_connection: WS handler
+    :param ed: existential deposit
+    :param chain_major_version: enum ChainMajorVersion
+    :param check_accounts_predicate: a pr
+    :param check_accounts_predicate_name:
+    :return:
+    """
+    accounts_that_do_not_meet_predicate = []
+    account_query = chain_connection.query_map('System', 'Account', page_size=1000)
```
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply scripts/accounts-invariants/pallet-balances-maintenance.py:162]: Fair point, but in our use case we must query them for the second time: once we do when fixiing balance/upgrading accounts, and this action changes state of the `System.Account`, so we must query those for the second time when we do account sanity checks.

File: scripts/accounts-invariants/pallet-balances-maintenance.py [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+    Prepare and send Balances.UpgradeAccounts call for all accounts on a chain
+    :param chain_connection: WS connection handler
+    :param input_args: script input arguments returned from argparse
+    :param ed: chain existential deposit
+    :param chain_major_version: enum ChainMajorVersion
+    :param sender_keypair: keypair of sender account
+    :return: None. Can raise exception in case of substrateinterface.SubstrateRequestException thrown
+    """
+    log.info("Querying all accounts.")
+    all_accounts_on_chain = list(map(lambda x: x[0], filter_false_accounts(chain_connection,
```
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply scripts/accounts-invariants/pallet-balances-maintenance.py:312]: See answer above

File: scripts/accounts-invariants/pallet-balances-maintenance.py [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+
+        extrinsic = chain_connection.create_signed_extrinsic(call=upgrade_accounts_call, keypair=sender_keypair)
+        log.info(
+            f"About to upgrade {len(account_ids_chunk)} accounts, each with "
+            f"{format_balance(chain_connection, existential_deposit)}")
+
+        submit_extrinsic(chain_connection, extrinsic, len(account_ids_chunk), args.dry_run)
+
+
+def chunks(list_of_elements, n):
```
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply scripts/accounts-invariants/pallet-balances-maintenance.py:335]: Yeah, but for now I assumed I'll stick to default Ubuntu-20.04 python system version that is 3.8.10.

File: scripts/accounts-invariants/pallet-balances-maintenance.py [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+                        action='store_true',
+                        help='Specify this switch if script should just print what if would do. Default: False')
+    parser.add_argument('--fix-free-balance',
+                        action='store_true',
+                        help='Specify this switch if script should find all accounts '
+                             'that have free < ED and send transfers so that free >= ED. '
+                             'It requires AlephNode < 12.X version and SENDER_ACCOUNT env to be set with '
+                             'a mnemonic phrase of sender account that has funds for transfers and fees. '
+                             'dust-accounts.json file is saved with all such accounts.'
+                             'Must be used exclusively with --upgrade-accounts. '
```
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply scripts/accounts-invariants/pallet-balances-maintenance.py:66]: Good point, done

File: scripts/accounts-invariants/pallet-balances-maintenance.py [hunk 5]
```diff
@@ -0,0 +1,451 @@
...
+                          chain_major_version,
+                          check_accounts_predicate,
+                          check_accounts_predicate_name=""):
+    """
+    Filters our accounts from the list on which predicate returns False
+    :param chain_connection: WS handler
+    :param ed: existential deposit
+    :param chain_major_version: enum ChainMajorVersion
+    :param check_accounts_predicate: a pr
+    :param check_accounts_predicate_name:
```
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply scripts/accounts-invariants/pallet-balances-maintenance.py]: Thanks, I got distracted meanwhile

File: scripts/accounts-invariants/pallet-balances-maintenance.py [hunk 6]
```diff
@@ -0,0 +1,451 @@
...
+    :return:
```
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply scripts/accounts-invariants/pallet-balances-maintenance.py]: Done

File: scripts/accounts-invariants/pallet-balances-maintenance.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kostekIV) [reply scripts/accounts-invariants/pallet-balances-maintenance.py:336]: sure

File: scripts/accounts-invariants/pallet-balances-maintenance.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kostekIV) [reply scripts/accounts-invariants/pallet-balances-maintenance.py:163]: Ok

**Review by kostekIV — APPROVED** [review]

**Pull request merged**